File: Client/Classes/GUI/GUIElement.lua
EGUI.Element = EGUI.Classlib.Inherit(EGUI.Object)
EGUI.Element.Type = "Unknown"

function EGUI.Element:Constructor(posX, posY, width, height, parent)
    self.m_Id = EGUI.Manager:Register(self.Type, self)

    self.m_PosX = posX
    self.m_PosY = posY
    self.m_Width = width
    self.m_Height = height
    self.m_Parent = parent

	self.m_Children = {}
	self.m_Visible = true
	self.m_Alpha = 1
	self.m_Alignment = "topLeft";

    self.m_Events = {}

	if self.m_Parent then
		self.m_Parent:AddChild(self)
	end
end

function EGUI.Element:Destructor()
	for index = #self.m_Children, -1 do
		local element = self.m_Children[index]
		EGUI.Classlib.Delete(element)
	end

    EGUI.Manager:CallEvent("GUIElementDestroy", self:GetId())
end

function EGUI.Element:GetId()
    return self.m_Id
end

function EGUI.Element:SetVisible(state)
	EGUI.Utils.CheckArgs("EGUI.Element:SetVisible", "boolean")
	self.m_Visible = state
    EGUI.Manager:CallEvent("GUIElementSetVisible", self:GetId(), self.m_Visible)
	return self
end

function EGUI.Element:IsVisible()
	return self.m_Visible
end

function EGUI.Element:SetAlpha(alpha)
	EGUI.Utils.CheckArgs("EGUI.Element:SetAlpha", "number")
	self.m_Alpha = alpha
    EGUI.Manager:CallEvent("GUIElementSetAlpha", self:GetId(), self.m_Alpha)
	return self
end

function EGUI.Element:GetAlpha()
	return self.m_Alpha
end

function EGUI.Element:SetPosition(x, y)
	EGUI.Utils.CheckArgs("EGUI.Element:SetPosition", "number", "number")
	self.m_PosX = x
	self.m_PosY = y
    EGUI.Manager:CallEvent("GUIElementSetPosition", self:GetId(), self.m_PosX, self.m_PosY)
	return self
end

function EGUI.Element:GetPosition()
	return self.m_PosX, self.m_PosY
end

function EGUI.Element:SetSize(width, height)
	EGUI.Utils.CheckArgs("EGUI.Element:SetSize", "number", "number")
	self.m_Width = width
	self.m_Height = height
    EGUI.Manager:CallEvent("GUIElementSetSize", self:GetId(), self.m_Width, self.m_Height)
	return self
end

function EGUI.Element:GetSize()
	return self.m_Width, self.m_Height
end

function EGUI.Element:SetAlignment(alignment)
	self.m_Alignment = alignment
    EGUI.Manager:CallEvent("GUIElementSetAlignment", self:GetId(), self.m_Alignment)
	return self
end

function EGUI.Element:GetAlignment()
	return self.m_Alignment
end

function EGUI.Element:AddChild(child)
	table.insert(self.m_Children, child)
end

function EGUI.Element:RemoveChild(child)
	for k, v in ipairs(self.m_Children) do
		if v == child then
			table.remove(self.m_Children, k)
			return
		end
	end
end

function EGUI.Element:HandleEvent(event, ...)

end

function EGUI.Element:Trigger(event, ...)
    if not self.m_Events[event] then return end

    for k, v in ipairs(self.m_Events[event]) do
        v(...)
    end
end

function EGUI.Element:Subscribe(event, handler)
    if not self.m_Events[event] then
        self.m_Events[event] = {}
    end

    table.insert(self.m_Events[event], handler)
end

function EGUI.Element:Unsubscribe(event, handler)
    if not self.m_Events[event] then
        self.m_Events[event] = {}
    end

    if not handler then
        self.m_Events[event] = {}
    else
        for k, v in ipairs(self.m_Events[event]) do
            if v == handler then
                table.remove(self.m_Events, k)
                return
            end
        end
    end
end

File: Client/UI/js/gui/GUIElement.js
class GUIElement {
    constructor(id, x, y, width, height, parent) {
        this.id = id;
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
        this.parent = parent;
		this.alignment = 'topLeft';
		/*
			topLeft
			topRight
			bottomLeft
			bottomRight
		*/

		this.visible = true;
		this.alpha = 1;

		this.children = [];

        guiManager.register(this.id, this);
    }

	destructor() {
        this.delete();
		guiManager.unregister(this.id);
	}

    render() {
        this.delete();

        this.baseDiv = document.createElement('div');
        this.baseDiv.setAttribute('id', this.id);

        if (this.parent) {
            const parent = guiManager.get(this.parent);

            if (parent) {
                parent.addChild(this);
            } else {
                document.body.append(this.baseDiv);
            }
        } else {
            document.body.append(this.baseDiv);
        }
    }

    delete() {
        if (this.baseDiv) {
            this.baseDiv.parentNode.removeChild(this.baseDiv);
            this.baseDiv = null;
        }
    }

	destroy() {
		for (var index = this.children.length; index >= 0; index--) {
			const element = this.children[index];

			if (element) {
				element.delete();
			}
		}

		this.delete();
	}

    update() {
        this.baseDiv.style.position = 'absolute';

		if (this.alignment === 'topLeft') {
			this.baseDiv.style.left = this.x + 'px';
			this.baseDiv.style.top = this.y + 'px';
			this.baseDiv.style.right = null;
			this.baseDiv.style.bottom = null;
		} else if (this.alignment === 'topRight') {
			this.baseDiv.style.left = null;
			this.baseDiv.style.top = this.y + 'px';
			this.baseDiv.style.right = this.x + 'px';
			this.baseDiv.style.bottom = null;
		} else if (this.alignment === 'bottomLeft') {
			this.baseDiv.style.left = this.x + 'px';
			this.baseDiv.style.top = null;
			this.baseDiv.style.right = null;
			this.baseDiv.style.bottom = this.y + 'px';
		} else if (this.alignment === 'bottomRight') {
			this.baseDiv.style.left = null;
			this.baseDiv.style.top = null;
			this.baseDiv.style.right = this.x + 'px';
			this.baseDiv.style.bottom = this.y + 'px';
		}

        this.baseDiv.style.width = this.width + 'px';
        this.baseDiv.style.height = this.height + 'px';

		this.baseDiv.style.display = this.visible ? 'block' : 'none';
		this.baseDiv.style.opacity = this.alpha;
    }

	setVisible(state) {
		this.visible = state;
		this.update();
	}

	setAlpha(alpha) {
		this.alpha = alpha;
		this.update();
	}

	setPosition(x, y) {
		this.x = x;
		this.y = y;
		this.update();
	}

	setSize(width, height) {
		this.width = width;
		this.height = height;
		this.update();
	}

	setAlignment(alignment) {
		this.alignment = alignment;
		this.update();
	}

    addChild(child) {
		this.children.push(child);
    }
}

EventsWrapper.Subscribe('GUIElementSetVisible', (id, state) => {
    const element = guiManager.get(id);

    if (element) {
        element.setVisible(state);
    }
});

EventsWrapper.Subscribe('GUIElementSetAlpha', (id, alpha) => {
    const element = guiManager.get(id);

    if (element) {
        element.setAlpha(alpha);
    }
});

EventsWrapper.Subscribe('GUIElementSetPosition', (id, x, y) => {
    const element = guiManager.get(id);

    if (element) {
        element.setPosition(x, y);
    }
});

EventsWrapper.Subscribe('GUIElementSetSize', (id, width, height) => {
    const element = guiManager.get(id);

    if (element) {
        element.setSize(width, height);
    }
});

EventsWrapper.Subscribe('GUIElementSetAlignment', (id, alignment) => {
    const element = guiManager.get(id);

    if (element) {
        element.setAlignment(alignment);
    }
});

EventsWrapper.Subscribe('GUIElementDestroy', (id) => {
    const element = guiManager.get(id);

    if (element) {
        element.destroy();
    }
});
